### ISSUE TITLE
Unable to specifiy relative urls with aiohttp router

### ISSUE THREAD

**Issue by jbcurtin** [aio-libs/aiohttp#946]
## Long story short

With running multipule reverse proxies via nginx, I would like to have the ability to have an application start from a relative endpoint rather than root. I am currently blocked by the frameworks control logic to force a root-level URL.
https://github.com/KeepSafe/aiohttp/blob/master/aiohttp/web_urldispatcher.py#L795
## Expected behaviour

Ideally, I would be allowed to execute three instances of aiohttp.web on three ports. This is already possible, but the ability to shard it is not. 

http://localhost/app-one/awesome.html
http://localhost/app-two/awesome.html
http://localhost/app-three/awesome.html

```
def init_app(argv):
    app = web.Application()
    app.router.add_route('GET', 'awesome.html', awesome_handler, name="awesome")
    return app
```
## Actual behaviour

```
File "lib/python3.5/site-packages/aiohttp/web.py", line 376, in <module>
    main(sys.argv[1:])
  File "lib/python3.5/site-packages/aiohttp/web.py", line 371, in main
    app = func(extra_argv)
  File "public_app.py", line 19, in init_app
    app.router.add_route('GET', 'authorize.html', authorize_handler, name="awesome")
  File "lib/python3.5/site-packages/aiohttp/web_urldispatcher.py", line 821, in add_route
    resource = self.add_resource(path, name=name)
  File "lib/python3.5/site-packages/aiohttp/web_urldispatcher.py", line 783, in add_resource
    raise ValueError("path should be started with /")
ValueError: path should be started with /
```
## Steps to reproduce

Execute the following:

```
#!/usr/bin/env python

from aiohttp import web

async def awesome_handler(request):
  return web.Response(body="<p>Awesome</p>".encode('utf-8'))

def init_app(argv):
  app = web.Application()
  app.router.add_route('GET', 'awesome.html', awesome_handler, name="awesome")
  return app
```

With the following commands:
python -m aiohttp.web -H 127.0.0.1 -P 8000 greetings.public_app:init_app
python -m aiohttp.web -H 127.0.0.1 -P 8001 greetings.public_app:init_app
python -m aiohttp.web -H 127.0.0.1 -P 8002 greetings.public_app:init_app
## Your environment

I'm running Ubuntu 14.04 LTS, with nginx sitting in front of the application as a reverse proxy.  In the upstream, the three hosts are configured with the respective ports. In the nginx.server context, a location with reverse_proxy. The context of this environment is out of scope, but for testing purposes, here is the configuration you can use.

```
upstream appone {
  server 127.0.0.1:8000 max_fails=2 fail_timeout=65;
}
upstream apptwo {
  server 127.0.0.1:8001 max_fails=2 fail_timeout=65;
}
upstream appthree {
  server 127.0.0.1:8002 max_fails=2 fail_timeout=65;
}
server {
  listen 80 default deferred;
  server_name localhost,127.0.0.1;
  location /application-one {
    include proxy_params;
    proxy_set_header  X-Forwarded-Host $host;
    proxy_set_header  X-Forwarded-Server $host;
    proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header  X-Real-IP $remote_addr;
    proxy_set_header  X-Referer $http_referer;
    proxy_set_header  Host $http_host;
    proxy_pass        http://app;
  }
  location /application-two {
    include proxy_params;
    proxy_set_header  X-Forwarded-Host $host;
    proxy_set_header  X-Forwarded-Server $host;
    proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header  X-Real-IP $remote_addr;
    proxy_set_header  X-Referer $http_referer;
    proxy_set_header  Host $http_host;
    proxy_pass        http://apptwo;
  }
  location /application-three {
    include proxy_params;
    proxy_set_header  X-Forwarded-Host $host;
    proxy_set_header  X-Forwarded-Server $host;
    proxy_set_header  X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header  X-Real-IP $remote_addr;
    proxy_set_header  X-Referer $http_referer;
    proxy_set_header  Host $http_host;
    proxy_pass        http://appthree;
  }
}
```

**Comment by popravich** [comment]
Hi,
Either use prefix in route:

``` python
app.router.add_route('GET', '/{appname}/awesome.html', awesome_handler, name="awesome")
```

or configure nginx, see proxy_pass + rewtite documentation

**Comment by jbcurtin** [comment]
Hi popravich, 

  Both of those suggested solutions do solve the immediate problem. However, they don't solve the full scope of the issue I've highlighted. If I were to move forward with either solution you've proposed, it would require me to employ code generation or 'dynamic' routing.

  The control structure, requiring the leading '/', is that from a design decision? If it is not a design decision, I can contribute code to fix it.

**Comment by popravich** [comment]
Requiring leading slash is by design.
Allowing relative routes will bring more uncertainty -- the urls would match by tail and it would become real problem with dynamic routes:

``` python
app.router.add_route('GET', 'awesome.html', awesome_handler, name='awesome')
# many other routes
app.router.add_route('GET', '/blog/{page}.html', blog_handler)
```

In this case any url ending with `awesome.html` will be handled by `awesome_handler`,
maybe it so awesome that it can handle and `awesome.html` but I am sure that one would
expect that `/blog/awesome.html` is handled by `blog_handler`.
Trying to solve this issue will make routing even more tangled...

**Comment by popravich** [comment]
Anyway, I think your problem can be solved in different way:
AFAIR, `UrlDispatcher` does not honor `SCRIPT_NAME` header and maybe we can make use of it.

**Comment by asvetlov** [comment]
@popravich I don't follow how `SCRIPT_NAME` may help but I totally agree with you: leading slash is design decision and nginx rewrite may solve request of @jbcurtin.

Closing the issue.
